keras_cv/layers/vit_layers.py

```diff
@@ -0,0 +1,204 @@
+    def __init__(self, patch_size, **kwargs):
+        super().__init__(**kwargs)
+        self.patch_size = patch_size
+
+    def call(self, images):
+        batch_size = tf.shape(images)[0]
+        patches = tf.image.extract_patches(
+            images=images,
+            sizes=[1, self.patch_size, self.patch_size, 1],
+            strides=[1, self.patch_size, self.patch_size, 1],
+            rates=[1, 1, 1, 1],
+            padding="VALID",
+        )
+        patch_dims = patches.shape[-1]
+        patches = tf.reshape(
+            patches, [batch_size, patches.shape[-2] * patches.shape[-2], patch_dims]
+        )
+        return patches
+
+    def get_config(self):
+        config = {
+            "patch_size": self.patch_size,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class PatchEmbedding(layers.Layer):
+    """
+    Layer to concat a class token, positionally embed and create a projection of patches made with the `Patching` layer
+    for Vision Transformers from:
+        - An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale
+        by Alexey Dosovitskiy et al. (https://arxiv.org/abs/2010.11929)
+
+    Based on Khalid Salama's implementation for:
+        - https://github.com/keras-team/keras-io/blob/master/examples/vision/image_classification_with_vision_transformer.py
+
+    args:
+        - project_dim: the dimensionality of the project_dim
+
+    Basic usage:
+
+    ```
+    patches = keras_cv.layers.Patching(patch_size)(batch_img)
+
+    project_dim = 1024
+    num_patches = patches.shape[1] # 196
+
+    encoded_patches = keras_cv.layers.PatchEmbedding(project_dim=project_dim)(patches)
+    print(encoded_patches.shape) # (1, 197, 1024)
+    ```
+    """
+
+    def __init__(self, project_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.project_dim = project_dim
+        self.linear_projection = layers.Dense(self.project_dim)
+
+    def build(self, input_shape):
+        self.class_token = tf.random.normal([1, 1, input_shape[-1]], name="class_token")
+        self.num_patches = input_shape[1]
+        self.position_embedding = layers.Embedding(
+            input_dim=self.num_patches + 1, output_dim=self.project_dim
+        )
+
+    def call(
+        self,
+        patch,
+        interpolate=False,
+        interpolate_width=None,
+        interpolate_height=None,
+        patch_size=None,
+    ):
+        # Add learnable class token before linear projection and positional embedding
+        patch_shape = tf.shape(patch)
+        class_token_broadcast = tf.cast(
+            tf.broadcast_to(
+                self.class_token,
+                [patch_shape[0], 1, patch_shape[-1]],
+            ),
+            dtype=patch.dtype,
+        )
+        patch = tf.concat([class_token_broadcast, patch], 1)
+        # num_patches + class token
```

Comment:
`# class token + num_patches`

Reply:
Whoops - left it in as a reminder during development as an explanation for the +1 in the next line. Removed!